packages/merge-styles/src/styleToClassName.ts

```diff
@@ -141,33 +141,86 @@ export function serializeRuleEntries(ruleEntries: { [key: string]: string | numb
   return allEntries.join('');
 }
 
-export function styleToClassName(...args: IStyle[]): string {
+export interface IRegistration {
+  className: string;
+  key: string;
+  args: IStyle[];
+  rulesToInsert: string[];
+}
+
+export function styleToRegistration(...args: IStyle[]): IRegistration | undefined {
   const rules: IRuleSet = extractRules(args);
   const key = getKeyForRules(rules);
-  let className = '';
 
   if (key) {
     const stylesheet = Stylesheet.getInstance();
+    const registration: Partial<IRegistration> = {
+      className: stylesheet.classNameFromKey(key),
+      key,
+      args
+    };
+
+    if (!registration.className) {
+      registration.className = stylesheet.getClassName(getDisplayName(rules));
+      const rulesToInsert: string[] = [];
+
+      for (const selector of rules.__order) {
+        rulesToInsert.push(
+          selector,
+          serializeRuleEntries(rules[selector])
+        );
+      }
+      registration.rulesToInsert = rulesToInsert;
+    }
 
-    className = stylesheet.classNameFromKey(key)!;
-
-    if (!className) {
-      className = stylesheet.getClassName(getDisplayName(rules));
-      const registeredRules: string[] = [];
+    return registration as IRegistration;
+  }
+}
 
-      for (let selector of rules.__order) {
-        const rulesToInsert: string = serializeRuleEntries(rules[selector]);
+export function applyRegistration(
+  registration: IRegistration,
+  classMap?: { [key: string]: string }
+): void {
+  const stylesheet = Stylesheet.getInstance();
+  const { className, key, args, rulesToInsert } = registration;
+
+  if (rulesToInsert) {
+    for (let i = 0; i < rulesToInsert.length; i += 2) {
+      const rules = rulesToInsert[i + 1];
+      if (rules) {
+        let selector = rulesToInsert[i];
+
+        // Fix selector using map.
+        selector = selector.replace(/(&)|\$(\w+)\b/g, (match: string, amp: string, cn: string): string => {
```

Comment:
This selector would only catch `$foo` of selector `$foo-bar`. And `.foo-bar` is a valid CSS class selector

Reply:
but it's not a valid javascript variable name. We're trying to produce a map that can be used in code which matches`I*ClassName` interface, right? Are you going to have an interface with 'foo-bar' as a member?

It's possible, but I didn't think it was worth optimizing for. If you disagree I can tweak the regex

Reply:
Honestly in looking at that line of code, i was more worried about creating a closure rather than a static function. I should fix that.

Reply:
Added support for dashes anyway. Just in case.